streamer/src/nonblocking/connection_rate_limiter.rs

```diff
@@ -0,0 +1,53 @@
+use {
+    governor::{DefaultDirectRateLimiter, DefaultKeyedRateLimiter, Quota, RateLimiter},
+    std::{net::IpAddr, num::NonZeroU32},
+};
+
+pub struct ConnectionRateLimiter {
+    limiter: DefaultKeyedRateLimiter<IpAddr>,
+}
+
+impl ConnectionRateLimiter {
+    pub fn new(limit_per_minute: u32) -> Self {
+        let quota = Quota::per_minute(NonZeroU32::new(limit_per_minute).unwrap()); // Adjust the rate limit as needed
+        Self {
+            limiter: DefaultKeyedRateLimiter::keyed(quota),
+        }
+    }
+
+    pub fn check(&self, ip: &IpAddr) -> bool {
+        // Acquire a permit from the rate limiter for the given IP address
+        if self.limiter.check_key(ip).is_ok() {
+            debug!("Request from IP {:?} allowed", ip);
+            true // Request allowed
+        } else {
+            debug!("Request from IP {:?} blocked", ip);
+            false // Request blocked
+        }
+    }
+}
+
+/// Connection rate limiter for enforcing connection rates from
+/// all clients.
+pub struct TotalConnectionRateLimiter {
```

Comment:
this replicode is unnecessary. it can just be a trait that we implement for each type. 

not too sure why the dependency doesn't already provide something like this. the APIs are very similar

Reply:
The two APIs are actually different. The TotalConnectionRateLimiter does not need the input to function, I changed the arg to simplify.

Reply:
Agreed that it's kind of tempting to try to parameterize this and turn it into a trait but tbh this is probably actually the most convenient thing to do.